### ISSUE TITLE
[Persistence] Evaluate Celestia's Sparse Map Tree implementation

### ISSUE THREAD

**Issue by Olshansk** [pokt-network/pocket#199]
## Objective

Determine if we can leverage an open source Sparse Merkle Tree implementation or if we need to implement our own.

## Origin Document

The implementation of Pocket Network's V1 state hash can be done independently of https://github.com/pokt-network/pocket/issues/147 by leveraging an existing library, but its correctness is dependent on that of the underlying library.

With cosmos leaving [IAVL](https://pkg.go.dev/github.com/cosmos/iavl), there are lots of other options available in the industry. https://github.com/cosmos/cosmos-sdk/issues/7100. Per the research done [in this deck](https://docs.google.com/presentation/d/1dgXVKLf_uqnviCMC-VfVsmy64j8ihY3AkySZDMr3tyU/edit#slide=id.g13a40e79564_0_58), Libra's [Jellyfish Merkle Tree](https://developers.diem.com/papers/jellyfish-merkle-tree/2021-01-14.pdf) seems to be the best path forward.

One of the best open source, fully in Go, implementations we've found online is from Celestia: https://github.com/celestiaorg/smt. I've reached out to see if this is being used in production via a [Tweet](https://twitter.com/olshansky/status/1555640151853989888) but have not heard back.

![Screen Shot 2022-09-08 at 6 22 47 PM](https://user-images.githubusercontent.com/1892194/189236334-dd41f807-7260-46df-936b-7fbbd1310892.png)

## Goals

- Review, evaluate & understand Celestia's open source SMT implementation
- **Stretch goal**: merge changes/ improvements upstream; opportunity to contribute back and work more closely together

## Deliverable

- [ ] A determination of the correction of Celestia's implementation
- [ ] A decision if this codebase could / should be used as our SMT implementation
- [ ] Optional: changes / improvements to Celestia's SMT

## Non-goals / Non-deliverables

- Implement any part of the Pocket Network V1 state hash

---

**Creator**: @Olshansk

**Comment by Olshansk** [comment]
Update. I sent out an initial PR with some questions related to their implementation & edits here: https://github.com/celestiaorg/smt/pull/74

I have a call with someone from their team this week as well

**Comment by Olshansk** [comment]
**tl;dr We will be using this Celestia's Merkle Tree for the time being, whichis being implemented in #285.**
---

### Future Work

Based on the research in [1] and [2], this fits our use case so far. Follow-up work (to be tracked in future tickets) to explain and verify this will include:
1. **Present**: Summarizing [1] and [2] and presenting it to the team (and community?) to provide a clear picture of the landscape of trees and why this was selected.
2. **Benchmark**: Using the tool in #285 to benchmark different KV stores when performance optimizations matter.
3. **Visualize**: Producitonizing the tools presented in https://github.com/celestiaorg/smt/pull/74 to improve observability & debugging in the tree.
4. **Export**: Building a tree export tool scoped in #340
5. Re-evaluate & verify this in Milestone M5 depending on how things and any new information that becomes available.

### Key Reasons - Technical

1. **Best available**: In short, to avoid IAVL (old, slow, complex), and while we're awaiting Verkle trees (too early), Sparse Merkle Tree implementations are the most efficient approach to generating a state hash today.
2. **State of Industry**: This implementation is a simplification of Libra's JMT, which is used by Aptos (formally Libra, formally Diem), optimized for high throughput and performance.
3. **Maintanble**: This implementation is done completely in Go and can therefore be forked, extended and maintained by the team.
4. **Configurable**: It is configurable for any key-value store rather than being tied to a specific one keeping forward-looking optionality.
5. **Swappable**: The interface is composed of 3 simple functions, allowing us to swap this out for other tree implementations in the future with minor "proxy interface" wrapper if need be also maintaining forward-looking optionality.

### Key Reasons - Technical Reassurance

These reassurances took place after the original decision as a result from [1] and [2] which is a big ➕ .

1. **Cosmos**: In [Cosmos' ADR 40](https://docs.cosmos.network/v0.46/architecture/adr-040-storage-and-smt-state-commitments.html), they were planning to migrate to using this SMT due to, in part, some of the technical reasons listed above.

![Screen Shot 2022-11-09 at 12 10 26 PM](https://user-images.githubusercontent.com/1892194/200931302-ec4857b5-6346-425e-afb5-2283e6c26cb8.png)

2. **Free upgrade**: A large refactor & performance improvement is currently [being reviewed](https://github.com/celestiaorg/smt/pull/73) to further improve this implementation, providing us with a "free lunch".

3. **Avoiding tech debt**: Seems like [1] is no longer going to be the case (at the time of writing) because Cosmos is too complex to adopt something anything other than IAVL (though that is the long-term goal given the discussion [here](https://github.com/cosmos/cosmos-sdk/issues/7100)), so instead, IAVL had to be forked and modified: https://github.com/celestiaorg/iavl/pull/8

![Screen Shot 2022-11-09 at 12 13 37 PM](https://user-images.githubusercontent.com/1892194/200933251-646c0322-098c-4f3d-88a6-7c57d647294a.png)

### References
[1] All the rough research notes on Trees & KV stores can be found [here](https://www.notion.so/pocketnetwork/Trees-Key-Value-Stores-a9801ab38e8441099795033f1e5b4928)
[2] The WIP presentation (at the time of writing) is available [here](https://docs.google.com/presentation/d/1hnoYbpMDroLAgfnUlOy8aqD3Q6A0jgWT_evqkI3-0fQ/edit#slide=id.p)

_Note: If you are an external contributor trying to access [1] or [2], reach out to the team on discord!_
